ssg/oval_object_model/oval_document.py

```diff
@@ -0,0 +1,307 @@
+        variable = load_variable(oval_variable_xml_element)
+        self._add_oval_component(variable, self.variables)
+
+    def get_xml_element(self):
+        root = self._get_oval_definition_el()
+        root.append(self._get_generator_el())
+        root.append(self._get_definitions_el())
+        root.append(self._get_tests_el())
+        root.append(self._get_objects_el())
+        if self.states:
+            root.append(self._get_states_el())
+        if self.variables:
+            root.append(self._get_variables_el())
+        return root
+
+    def save_as_xml(self, fd):
+        root = self.get_xml_element()
+        ElementTree.indent(root, space=" ", level=0)
+        ElementTree.ElementTree(root).write(fd, xml_declaration=True, encoding="utf-8")
+
+    def _get_component_el(self, tag, values):
+        xml_el = ElementTree.Element("{%s}%s" % (OVAL_NAMESPACES.get("oval_def"), tag))
+        for val in values:
+            xml_el.append(val.get_xml_element())
+        return xml_el
+
+    def _get_definitions_el(self):
+        return self._get_component_el("definitions", self.definitions.values())
+
+    def _get_tests_el(self):
+        return self._get_component_el("tests", self.tests.values())
+
+    def _get_objects_el(self):
+        return self._get_component_el("objects", self.objects.values())
+
+    def _get_states_el(self):
+        return self._get_component_el("states", self.states.values())
+
+    def _get_variables_el(self):
+        return self._get_component_el("variables", self.variables.values())
+
+    def _get_generator_el(self):
+        generator_el = ElementTree.Element(
+            "{%s}generator" % OVAL_NAMESPACES.get("oval_def")
+        )
+
+        product_name_el = ElementTree.Element(
+            "{%s}product_name" % OVAL_NAMESPACES.get("oval")
+        )
+        product_name_el.text = self.oval_header_info.get("product_name")
+
+        generator_el.append(product_name_el)
+
+        product_version_el = ElementTree.Element(
+            "{%s}product_version" % OVAL_NAMESPACES.get("oval")
+        )
+        product_version_el.text = self.oval_header_info.get("product_version")
+
+        generator_el.append(product_version_el)
+
+        schema_version_el = ElementTree.Element(
+            "{%s}schema_version" % OVAL_NAMESPACES.get("oval")
+        )
+        schema_version_el.text = str(self.oval_header_info.get("schema_version"))
+        generator_el.append(schema_version_el)
+
+        timestamp_el = ElementTree.Element(
+            "{%s}timestamp" % OVAL_NAMESPACES.get("oval")
+        )
+        timestamp_el.text = str(timestamp)
+        generator_el.append(timestamp_el)
+
+        return generator_el
+
+    def _get_oval_definition_el(self):
+        oval_definition_el = ElementTree.Element(
+            "{%s}oval_definitions" % OVAL_NAMESPACES.get("oval_def")
+        )
+
+        space = "         "
+        oval_definition_el.set(
```

Comment:
This looks really suspicious. The library should declare the namespaces and prefixes automatically without you touching it.

Reply:
In this case, a special attribute with links to schemas is created. This attribute is intended for the validator.  For this purpose, the namespace and attributes are set manually.

Reply:
Could you give an example of the special attribute? I think we need to take a detailed look into this.

Reply:
The `xsi:schemaLocation` attribute is with a namespace attribute. I figured out how to add this namespace attribute correctly. It's nothing special. But in the OVAL context, it's different.